### PR TITLE
Advanced Spray Color

### LINKED ISSUE #248: [ENHANCEMENT] Random color option for spray logo
title

### PR DESCRIPTION
**Pull request by kai-li-wop**
This should improve and extend the spray color options for Spray Your Color game types.

- [x] Extend spray color selection from 6 to the full game color table of 8, adding cyan and magenta as spray colors
- [x] Extend player settings menu to handle all 8 colors and adjust the player settings menu background texture
- [x] Extend select logo ingame menu to handle all 8 colors
- [x] Add a random color option to the player settings menu to allow the user to randomize the spray color and injector trail color with each shot (see #248)
- [x] Add random color to user config string
- [x] Implement a random color function for Injector trail and flash
- [x] Implement a random color function for Spraypistol trail and spray logo / spray mark (for not hitting a spray wall)
- [x] New spray mark for missing the wall (Easter Egg)

### CODE REVIEW

**Comment by kai-li-wop** [comment]
For some reason, the spray logo and spray pistol trail have a different random color while kma trail and flash have the same random color. I don't know why.

File: code/cgame/cg_spraylogo.c [hunk 1]
```diff
@@ -474,10 +474,7 @@ void Add_LogoToDrawList(const vec3_t origin, vec3_t dir, qhandle_t shader, float
 		lp->spraytime = cg.time;
 		lp->logoShader = shader;
 		lp->numVerts = lf->numPoints;
-		lp->color[0] = (float)color[0] / 255.0f;
-		lp->color[1] = (float)color[1] / 255.0f;
-		lp->color[2] = (float)color[2] / 255.0f;
-		lp->color[3] = (float)color[3] / 255.0f;
+		Vector4Copy(color, lp->color);
```
Comment (mgerhardy) [inline code/cgame/cg_spraylogo.c]: this won't work - the color array here is of type byte, not float - and as such the values are in the range 0-255, for float color values they are normalized in the range 0.0-1.0. The dividing here is still needed.

File: code/game/g_bot.c [hunk 2]
```diff
@@ -554,17 +554,10 @@ static void G_AddBot(const char *name, float skill, const char *team, int delay,
...
 	if (!*s) {
-		s = "4";
-	}
-	Info_SetValueForKey(userinfo, key, s);
-
-	key = "color2";
-	s = Info_ValueForKey(botinfo, key);
-	if (!*s) {
-		s = "5";
+		s = "2";
```
Comment (mgerhardy) [inline code/game/g_bot.c]: comment for COLOR_GREEN would be nice.

`s = "2"; // COLOR_GREEN`

File: code/cgame/cg_weapons.c [hunk 3]
```diff
@@ -1696,6 +1710,14 @@ void CG_FireWeapon(centity_t *cent) {
 		}
 	}
 
+	// random spray color if spray pistol or injector is used
+	// and random color menu option is enabled
+	if (cgs.clientinfo[cent->currentState.clientNum].randomcolor == 1 &&
+			(ent->weapon == WP_KMA97 || ent->weapon == WP_SPRAYPISTOL)) {
+		int rnd = (int)(random() * ((float)NUM_SPRAYCOLORS - 0.1f));
```
Comment (mgerhardy) [inline code/cgame/cg_weapons.c]: here you can use `randomindex(NUM_SPRAYCOLORS)`

File: code/game/g_client.c [hunk 4]
```diff
@@ -850,26 +849,27 @@ void ClientUserinfoChanged(int clientNum) {
...
 	// print scoreboards, display models, and play custom sounds
 	if (ent->r.svFlags & SVF_BOT) {
 		// cyr{
-		char *rnd_str;
-		int rnd;
-		rnd = random() * 5.9;
-		rnd_str = va("%d", rnd);
-		s = va("n\\%s\\t\\%i\\model\\%s\\hmodel\\%s\\c1\\%s\\c2\\%s\\hc\\%i\\w\\%i\\l\\%i\\skill\\%s\\tl\\%d"
+		char *rndSprayColor;
+		int rnd = random() * ((float)NUM_SPRAYCOLORS - 0.1f);
```
Comment (mgerhardy) [inline code/game/g_client.c]: here you can use `randomindex(NUM_SPRAYCOLORS)`

File: code/game/g_client.c [hunk 5]
```diff
@@ -850,26 +849,27 @@ void ClientUserinfoChanged(int clientNum) {
...
 	// send over a subset of the userinfo keys so other clients can
 	// print scoreboards, display models, and play custom sounds
 	if (ent->r.svFlags & SVF_BOT) {
 		// cyr{
-		char *rnd_str;
-		int rnd;
-		rnd = random() * 5.9;
-		rnd_str = va("%d", rnd);
-		s = va("n\\%s\\t\\%i\\model\\%s\\hmodel\\%s\\c1\\%s\\c2\\%s\\hc\\%i\\w\\%i\\l\\%i\\skill\\%s\\tl\\%d"
+		char *rndSprayColor;
```
Comment (mgerhardy) [inline code/game/g_client.c]: this should be `const char *`

File: code/qcommon/q_shared.h [hunk 6]
```diff
@@ -445,9 +445,7 @@ qboolean Q_IsColorString(const char *p); // ^[0-9a-zA-Z]
 #define S_COLOR_WHITE "^7"
 
 extern vec4_t g_color_table[8];
-extern vec4_t spraycolors[];
-//#define NUM_SPRAYCOLORS ( sizeof( spraycolors[] ) / sizeof( spraycolors[0] ) )
-#define NUM_SPRAYCOLORS 6
+#define NUM_SPRAYCOLORS 8
```
Comment (mgerhardy) [inline code/qcommon/q_shared.h]: maybe move this define over `g_color_table` and use `extern vec4_t g_color_table[NUM_SPRAYCOLORS];` (and maybe even rename `NUM_SPRAYCOLORS` to `NUM_COLORS`)

File: code/cgame/cg_spraylogo.c (hunk 1 above)
Comment (mgerhardy) [reply code/cgame/cg_spraylogo.c]: ah you also changed from byte to float - i missed that type change...

File: code/cgame/cg_weapons.c (hunk 3 above)
Comment (kai-li-wop) [reply code/cgame/cg_weapons.c:1717]: randomindex(NUM_SPRAYCOLORS -1) I think.

File: code/cgame/cg_spraylogo.c [hunk 7]
```diff
@@ -585,12 +582,12 @@ void ActiveChooseLogoMenu(void) {
 		else
 			spraycolor = colorBlue;
 	} else {
-		trap_Cvar_VariableStringBuffer("syc_color", colorchar, sizeof(colorchar));
+		trap_Cvar_VariableStringBuffer("spraycolor", colorchar, sizeof(colorchar));
 
-		if (colorchar[0] > '5' || colorchar[0] < '0')
+		if (colorchar[0] > '7' || colorchar[0] < '0')
```
Comment (mgerhardy) [inline code/cgame/cg_spraylogo.c]: should be replaced by the color constants

File: code/cgame/cg_spraylogo.c [hunk 8]
```diff
@@ -585,12 +582,12 @@ void ActiveChooseLogoMenu(void) {
...
 			colorchar[0] = '0';
 
-		spraycolor = spraycolors[(int)(colorchar[0] - '0')];
+		spraycolor = g_color_table[(int)(colorchar[0] - '0')];
```
Comment (mgerhardy) [inline code/cgame/cg_spraylogo.c]: use ColorIndex(colorchar[0]) here

**Comment by mgerhardy** [comment]
```sh
\g_gametype 3
\devmap wop_jail
\set randomcolor 1
\give ammo_spray
\give ammo_spray
\give ammo_spray
\give ammo_spray
\give ammo_spray
\give ammo_spray
\give ammo_spray
\give ammo_spray
\setviewpos steck1
```

**Comment by 3aTmE** [comment]
Nicely done! 
Everything seems to be working well. 

**What was tested ?**
- Random Color Option for the Spraylogo / Injector beam
- Easter Egg
- The New Colors for the Spray Logo / Injector beam
- Beeing able to select the new colors in the Player menu

**Pull request merged**